Modify each client's QoS parameter applied time from client_rec creation time to each request's tagging creation time.

bspark8 opened the pull request:
The dmclock library updates each client's QoS parameter information when the client related class is created, which is fixed during system online time. it caused the inconvenience of restarting the system when the user want to change the QoS parameter.
Thus I modified each client's QoS parameter applied time from client related class creation time to each request's tagging creation time to support each client's online QoS parameter changes.

ivancich commented:
I can see why this would be helpful. We don't know if the call to client_info_f might be expensive, but if it is, it could be addressed by a later modification to call to see if the parameters have changed and only calling client_info_f if it's aware of changes.

Why is the information stored in the RequestTag (reservation_inv) even if it's also stored in the ClientRec (popped_resv_inv)? Could the ClientRec's value be a previous value if there was a change?

bspark8 commented:
First, yes, you're right. At first, I considered the notification method as you said.
However, in some cases, I realized that the dynamic calling method is somewhat simpler and lighter than we thought.

For example, I will use the dmclock library in Ceph.
In case of per-pool dmclock parameters, dmclock parameters will be stored at OSD map which includes the each pool information. when certain pool's QoS parameters are changed, OSD map will be updated with the new epoch and the each OSD which has older OSD map will try to take new OSD map. And the OSD will try to call ClientRec update procedure through dmclock library. But right then, maybe all existing ClientRecs will be updated because the OSD can't get certain pool's ID from OSD map update event.
To handle above case, quite amount of Ceph codes will be modified.
Inserting ClientRec update procedure to all OSD map updated cases also need quite a bit of Ceph code modification.

For generalization, it would be nice to provide both dynamic calling & notification method in dmclock library.

Second is about reservation_inv.
That reason is simple. When we use the dynamic calling method, as you said, the cost of client_info_f can be expensive. Thus to save a little cost, I'd like to avoid the expensive floating point division.
For example, If we remove reservation_inv from the RequestTag, the code will be changed from "top.set_popped_resv_inv(first.tag.reservation_inv);" to "top.set_popped_resv_inv(0.0 == first.tag.reservation ? 0.0 : 1.0 / first.tag.reservation)".
But, if that does not help performance, it would be better to remove reservation_inv from the RequestTag.

bspark8 commented:
I have added the work you recommended.

First, I added option DO_DYNAMIC_CLI_INFO_F.
If the DO_DYNAMIC_CLI_INFO_F is explicitly defined, the dmclock library the user can call client_info_f function dynamically when creating request's tagging.
It will help the user to use dmclock library with flexibility.
The default option is the original (previous) mode.

Second, I added update_client_info API in order to support the notification method.
This allow user to update the certain ClientRec's ClientInfo at the original (previous) mode.

ivancich commented on src/dmclock_server.h:
```diff
@@ -1084,6 +1114,24 @@ namespace crimson {
+	auto client_it = client_map.find(client_id);
+	if (client_map.end() != client_it) {
+	  ClientRec& client = (*client_it->second);
+	  client.info = client_info_f(client_id);
```
Has this code been tested? ClientInfo declares all its variables as const, and it has no assignment operator. If this is to work, I think we'll have to change ClientInfo so the variables are no longer const and there is an assignment operator.

bspark8 replied:
Great catch! Thank you.

ivancich commented on src/dmclock_server.h:
```diff
@@ -1084,6 +1114,24 @@ namespace crimson {
+      // data_mtx must be held by caller
+      void do_update_client_infos() {
+	for (auto i : client_map) {
+	  i.second->info = client_info_f(i.second->client);
```
I think this will have the same issues outlined above with do_update_client_info.

ivancich commented on src/dmclock_server.h:
```diff
@@ -1292,6 +1340,18 @@ namespace crimson {
       } // pull_request
 
 
+      void update_client_info(const C& client_id) {
```
This function is repeated for both PullPriorityQueue and PushPriorityQueue. Shouldn't this be a public function in PriorityQueueBase?

bspark8 replied:
I removed repeated update_client_info(s) in both PullPriorityQueue and PushPriorityQueue.

ivancich commented on src/dmclock_server.h:
```diff
@@ -1292,6 +1340,18 @@ namespace crimson {
+      }
+
+
+      void update_client_info() {
```
This function is repeated for both PullPriorityQueue and PushPriorityQueue. Shouldn't this be a public function in PriorityQueueBase?

ivancich commented on src/dmclock_server.h:
```diff
@@ -1453,6 +1513,18 @@ namespace crimson {
       }
 
+
+      void update_client_info(const C& client_id) {
```
This function is repeated for both PullPriorityQueue and PushPriorityQueue. Shouldn't this be a public function in PriorityQueueBase?

ivancich commented on src/dmclock_server.h:
```diff
@@ -1453,6 +1513,18 @@ namespace crimson {
+      }
+
+
+      void update_client_info() {
```
This function is repeated for both PullPriorityQueue and PushPriorityQueue. Shouldn't this be a public function in PriorityQueueBase?

ivancich commented on src/dmclock_server.h:
```diff
@@ -821,14 +831,23 @@ namespace crimson {
 	if (!client.has_request()) {
-	  tag = RequestTag(client.get_req_tag(), client.info,
-			   req_params, time, cost);
+#ifndef DO_DYNAMIC_CLI_INFO_F
```
Looking at how to test this thoroughly along with the existing DO_NOT_DELAY_TAG_CALC, since they interact, you'd have to build the code in four different ways with each combination of the settings. For that reason I'm wondering whether if DO_DYNAMIC_CLI_INFO_F would be better as a boolean template parameter to the PriorityQueue classes. Would you please explore that? You don't have to do anything with DO_NOT_DELAY_TAG_CALC, as if you can successfully change DO_DYNAMIC_CLI_INFO_F to a template parameter, I might then make an effort to do the same with DO_NOT_DELAY_TAG_CALC.

bspark8 replied:
As you recommended, I removed DO_DYNAMIC_CLI_INFO_F and used the boolean template parameter to the PriorityQueueBase class.

bspark8 commented:
Thank you for your feedback!
I completed rebase & code rearrangement and checked a full unit test pass for the four cases you mentioned.
In addition, I added dmclock_server_pull.update_client_info unit test for update_client_info.
Also, for dynamic_cli_info_f, I added dmclock_server_pull.dynamic_cli_info_f unit test.

ivancich commented on src/dmclock_server.h:
```diff
@@ -744,6 +764,14 @@ namespace crimson {
 
 
+      inline const ClientInfo get_cli_info(ClientRec& client) const {
+	if (dynamic_cli_info_f) {
```
Since this is the only use of the variable `dynamic_cli_info_f`, and I assume you like a well-named variable here rather than `U1`, why not put this in the line before
`constexpr bool dynamic_cli_info_f = U1;`
and then use `dynamic_cli_info_f` in your conditional?

bspark8 replied:
I had originally intended to use well-named variable to support complie time constant, as you said.
The reason I used bool variable was to change the use of dynamic_cli_info_f via the configuration change even after the queue was created.
However, it seems that it is not necessary now.

ivancich commented on src/dmclock_server.h:
```diff
@@ -696,6 +715,7 @@ namespace crimson {
       bool             allow_limit_break;
 
       std::atomic_bool finishing;
+      bool             dynamic_cli_info_f = U1;
```
We can remove this variable and still get the use of a well-named variable by using a `constexpr` in your `get_cli_info function` below.

ivancich commented on test/test_dmclock_server.cc:
```diff
@@ -639,6 +639,179 @@ namespace crimson {
+
+      QueueRef pq;
+
+      auto client_info_f = [&] (ClientId c) -> dmc::ClientInfo {
```
To test the dynamic nature of the client_info function, wouldn't we need this function to return a different ClientInfo object for the same client at different points in the run? In other words at some point you might want to change the values of, say, info1, and see if that affects later pulls from the queue.

bspark8 replied:
Thank you and to test the dynamic nature of the client_info function, I modified the function client_info_f according to your opinion.

ivancich commented on test/test_dmclock_server.cc:
```diff
@@ -639,6 +639,179 @@ namespace crimson {
     } // dmclock_server_pull.pull_reservation
 
 
+    TEST(dmclock_server_pull, update_client_info) {
```
This test has failures when `DO_NOT_DELAY_TAG_CALC` is defined.

bspark8 replied:
Sorry for the lack of check. I was fixed the failure situation.

ivancich commented on test/test_dmclock_server.cc:
```diff
@@ -639,6 +639,179 @@ namespace crimson {
+    }
+
+
+    TEST(dmclock_server_pull, dynamic_cli_info_f) {
```
This test has failures when `DO_NOT_DELAY_TAG_CALC` is defined.

bspark8 commented:
I have left messages and completed the changes requested. Thank you!

ivancich commented:
Thank you!
